src/components/DataTable/DataTable-story.js

```diff
@@ -0,0 +1,580 @@
+          : aContent[key];
+      const varB =
+        typeof bContent[key] === 'string'
+          ? bContent[key].toUpperCase()
+          : bContent[key];
+
+      let comparison = 0;
+      if (varA > varB) {
+        comparison = 1;
+      } else if (varA < varB) {
+        comparison = -1;
+      }
+
+      return order === 'desc' ? comparison * -1 : comparison;
+    };
+  };
+
+  sortRow = (query, dir) => {
+    const newRows = this.state.rows.sort(this.compareVals(query, dir));
+    this.setState({
+      rows: newRows,
+    });
+  };
+
+  searchTable = evt => {
+    let newRows = [];
+    this.state.rows.map(obj => {
+      Object.keys(obj).forEach(key => {
+        if (
+          obj[key]
+            .toUpperCase()
+            .includes(evt.currentTarget.value.toUpperCase()) &&
+          !newRows.includes(obj)
+        ) {
+          newRows.push(obj);
+        }
+      });
+    });
+    this.setState({
+      filteredRows: newRows,
+    });
+  };
+
+  render() {
+    const checkedItems = this.getCheckedItems();
+    const showBatchActions = checkedItems > 0;
+    const tableClasses = classNames({
+      'bx--data-table-v2': true,
+      'bx--data-table-v2--zebra': this.props.zebra,
+    });
+    return (
+      <div>
+        <DataTableContainer title="Table title">
+          <DataTableToolbar>
+            <DataTableBatchActions
+              totalSelected={checkedItems}
+              showBatchActions={showBatchActions}
+              handleClick={this.clearAll}>
+              <DataTableActionList>
+                <DataTableBatchAction onClick={action('Batch Action 1')}>
+                  Ghost
+                </DataTableBatchAction>
+                <DataTableBatchAction onClick={action('Batch Action 2')}>
+                  Ghost
+                </DataTableBatchAction>
+                <DataTableBatchAction onClick={action('Batch Action 3')}>
+                  Ghost
+                </DataTableBatchAction>
+              </DataTableActionList>
+            </DataTableBatchActions>
+            <DataTableSearch
+              onInput={this.searchTable}
+              onChange={this.searchTable}
+            />
+            <DataTableToolbarContent>
+              <DataTableToolbarAction
+                iconName="download"
+                iconDescription="Download"
+                onClick={action('Toolbar Action 1')}
+              />
+              <DataTableToolbarAction
+                iconName="edit"
+                iconDescription="Edit"
+                onClick={action('Toolbar Action 1')}
+              />
+              <DataTableToolbarAction
+                iconName="settings"
+                iconDescription="Settings"
+                onClick={action('Toolbar Action 1')}
+              />
+              <Button onClick={action('Add new row')} kind="primary">
+                Add new
+              </Button>
+            </DataTableToolbarContent>
+          </DataTableToolbar>
+          <DataTable
```

Comment:
What if `<DataTable>` looked something like:

```js
<DataTable
  initialRows={data.rows}
  headers={['Foo', 'Bar', 'Baz']}
  {/* enable checkboxes, name can definitely change! */}
  selectable={true}
  render={({
    // Current page to render
    page,

    // Headers passed through from above
    headers,

    // Prop getters for children components to keep in sync with internal
    // state in DataTable
    getPaginationProps,
    // Returns things like `onClick` handlers if the item is select-able
    getSelectionProps,
    // Potentially useful for inline editing
    getColumnProps,
    getHeaderProps,

    // Sorters
    sortBy,
    sortByFoo,
    sortByBar,
    sortByBaz,

    // Selectors
    selectRow,
  }) => (
    <table>
      <TableHead>
        {headers.map(header => (
          <TableHeader {...getHeaderProps({ header })}>
            {header}
          </TableHeader>
        ))}
      </TableHead>
      <TableBody>
        {page.rows.map(row => (
          <TableRow key={`row-select-${row.id}`}>
            {row.isSelectable && (
              <TableData {...getSelectionProps()} />
            )}
          </TableRow>
          {row.columns.map(column => (
            <TableData {...getColumnProps({ column })} />
          ))}
        ))}
      </TableBody>
    </table>
  )}
/>
```

Reply:
Hmm, alright. Though I think in this case, we shouldn't call it `TableData` but `TableCell` to be consistent with the html table concept.

As long as initialRows is optional and we can potentially bypass the paging entirely (for the sake of infinite scroll). 

The sorters are interesting though, are they triggering sort or are they performing the sort? I thought sort itself might sit outside this component, since it might be triggered from outside the component (say, from a url or redux action). Same concern for selectRow. I'm not sure how much the the table itself should track those properties, since they can be triggered by many external sources.

Reply:
I agree about the sorting part, I think having that be controlled outside the component is probably best. Having `sortBy`, `sortByFoo`, ... , seems a bit redundant as there might be a large number of queries users would want to sort by.

Reply:
I'm a bit confused about this suggested setup. I think the header part is not customizable enough this way, as I think the user should be able to specify what they want to sort each column by - I might be missing something though. I also feel like I should wait with the pagination parts until after the v1 release - since pagination is not part of the actual component right now? I might just need some help visualizing what the `DataTable` component would look like, it's not apparently clear to me with this pattern.

Reply:
Also I think `DataTableData` is fine as w3 explains the `td` element as such:
`Use the HTML <td> element to define a table data`.

Reply:
Huh , look at that. Fair enough. I always think of them as cells. It doesn't help that "DataTableData" doesn't exactly roll off the tongue I guess. Data always feels like such an overloaded term.

Reply:
Whoops, duh, the `sortBy` stuff was in an earlier draft I had as the comment which I edited/replaced. `getHeaderProps` would include the sort functionality, potentially, or they could manually put the `sortBy` field on it if they wanted and did `sortBy(header)` and it would set it up for them dynamically.

Reply:
Link to an example of what I was thinking: https://codesandbox.io/s/pnzk98jq